File: kernel/src/mouse.zig
const gfx = @import("gfx.zig");
const io = @import("io.zig");
const isr = @import("isr.zig");

const MouseState = enum {
    mouse_state_0,
    mouse_state_1,
    mouse_state_2,
};

const Buttons = struct {
    left: bool,
    middle: bool,
    right: bool,
};

var state = MouseState.mouse_state_0;
var byte_0: u8 = undefined;
var byte_1: u8 = undefined;
var byte_2: u8 = undefined;

pub var coordinates = gfx.Point{ .x = 0, .y = 0 };
pub var buttons = Buttons{ .left = false, .middle = false, .right = false };

pub fn initialize() void {
    io.outb(0x64, 0xAD);
    mouse_wait(false);
    io.outb(0x64, 0xA8);
    mouse_wait(false);

    io.outb(0x64, 0xD4);
    mouse_wait(false);
    io.outb(0x60, 0xF6);
    mouse_wait(true);
    _ = io.inb(0x60);

    mouse_wait(false);
    io.outb(0x64, 0x20);
    mouse_wait(true);
    var status = io.inb(0x60);
    status |= 1 << 1;
    status &= ~@as(u8, 1 << 5);
    mouse_wait(false);
    io.outb(0x64, 0x60);
    mouse_wait(false);
    io.outb(0x60, status);
    mouse_wait(false);
    io.outb(0x64, 0xD4);
    mouse_wait(false);
    io.outb(0x60, 0xF4);
    mouse_wait(true);
    _ = io.inb(0x60);
    mouse_wait(false);
    io.outb(0x64, 0xAE);
    _ = io.inb(0x60);
    _ = io.inb(0x60);
    _ = io.inb(0x60);
    _ = io.inb(0x60);

    isr.install_handler(12, interrupt_handler);
}

pub fn interrupt_handler() void {
    if (io.inb(0x64) & 1 == 0) return;
    switch (state) {
        .mouse_state_0 => {
            state = MouseState.mouse_state_1;
            byte_0 = io.inb(0x60);
        },
        .mouse_state_1 => {
            state = MouseState.mouse_state_2;
            byte_1 = io.inb(0x60);
        },
        .mouse_state_2 => {
            state = MouseState.mouse_state_0;
            byte_2 = io.inb(0x60);
            update();
        },
    }
}

fn update() void {
    const rel_x: i8 = @truncate(byte_1 - (@as(i16, ((byte_0 << 4))) & 0x100));
    const rel_y: i8 = @truncate(byte_2 - (@as(i16, ((byte_0 << 3))) & 0x100));
    if (rel_x > 0) coordinates.x += @intCast(rel_x) else if (coordinates.x > -rel_x) coordinates.x -= @intCast(-rel_x);
    if (rel_y > 0) {
        if (coordinates.y > rel_y) coordinates.y -= @intCast(rel_y);
    } else coordinates.y += @intCast(-rel_y);
    if (coordinates.x > 640) coordinates.x = 640;
    if (coordinates.y > 480) coordinates.y = 480;
    buttons.left = byte_0 & 1 != 0;
    buttons.middle = byte_0 & 4 != 0;
    buttons.right = byte_0 & 2 != 0;
}

fn mouse_wait(read: bool) void {
    var timeout: u32 = 100000;
    if (read) {
        while (timeout != 0) {
            if ((io.inb(0x64) & 1) != 0)
                return;
            timeout -= 1;
        }
    } else {
        while (timeout != 0) {
            if ((io.inb(0x64) & 2) != 0)
                return;
            timeout -= 1;
        }
    }
}

File: kernel/mouse.c
#include <kernel/mouse.h>
#include <kernel/isr.h>
#include <kernel/io.h>
#include <kernel/trapframe.h>
#include <kernel/event.h>

#include <stdint.h>
#include <stdbool.h>

uint32_t state = mouse_state_0;
uint8_t byte_0;
uint8_t byte_1;
uint8_t byte_2;

mouse_buttons_t mouse_buttons = { .left = false, .middle = false, .right = false };
uint16_t mouse_x = 0;
uint16_t mouse_y = 0;

static void update_mouse() {
    int16_t rel_x = byte_1 - ((byte_0 << 4) & 0x100);
    int16_t rel_y = byte_2 - ((byte_0 << 3) & 0x100);
    if (rel_x > 0) mouse_x += rel_x; else if ((int16_t) mouse_x > -rel_x) mouse_x -= (-rel_x);
    if (rel_y > 0) {
        if (mouse_y > rel_y) mouse_y -= rel_y;
    } else mouse_y += (-rel_y);
    if (mouse_x > 640) mouse_x = 640;
    if (mouse_y > 480) mouse_y = 480;

    mouse_buttons_t old_mouse_buttons;
    old_mouse_buttons.left = mouse_buttons.left;
    old_mouse_buttons.middle = mouse_buttons.middle;
    old_mouse_buttons.right = mouse_buttons.right;

    mouse_buttons.left = (byte_0 & 1) != 0;
    mouse_buttons.middle = (byte_0 & 4) != 0;
    mouse_buttons.right = (byte_0 & 2) != 0;

    // TODO: other buttons?
    if (mouse_buttons.left != old_mouse_buttons.left) {
        event_t event;
        event.type = mouse_buttons.left ? MOUSE_DOWN : MOUSE_UP;
        event.arg0 = mouse_x;
        event.arg1 = mouse_y;
        new_event(&event);
    }
}

void init_mouse() {
    outb(0x64, 0x20);
    uint8_t status = inb(0x60);
    status |= 1 << 1;
    status &= ~(1 << 5);
    outb(0x64, 0x60);
    while ((inb(0x64) & 1) != 0);
    outb(0x60, status);
    outb(0x64, 0xD4);
    while ((inb(0x64) & 1) != 0);
    outb(0x60, 0xF4);
    inb(0x60);
    install_interrupt_handler(12, mouse_interrupt_handler);
}

void mouse_interrupt_handler(uint8_t irq, trap_frame_t *trap_frame, uint32_t error) {
    (void) irq;
    (void) trap_frame;
    (void) error;
    if ((inb(0x64) & 1) == 0) return;
    switch (state) {
        case mouse_state_0: {
            state = mouse_state_1;
            byte_0 = inb(0x60);
            break;
        }
        case mouse_state_1: {
            state = mouse_state_2;
            byte_1 = inb(0x60);
            break;
        }
        case mouse_state_2: {
            state = mouse_state_0;
            byte_2 = inb(0x60);
            update_mouse();
            break;
        }
    }
}
